### PR TITLE
Fix iterator types for embedding bag sum mode

### PR DESCRIPTION
**Pull request by qedawkins**

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/IndirectDataMovement.cpp [hunk 1]
```diff
@@ -309,8 +309,12 @@ class ConvertAtenEmbeddingBagPaddingIdxOp
           outputIndexingMap,
     };
 
-    SmallVector<StringRef> iteratorTypes(iterationMapDimension,
+    // This gives the correct iterator types for sum mode
+    SmallVector<StringRef> iteratorTypes(indicesTy.getRank(),
                                            getParallelIteratorTypeName());
+    SmallVector<StringRef> reductionIteratorTypes(weightTy.getRank(),
+                                           getReductionIteratorTypeName());
+    iteratorTypes.insert(iteratorTypes.end(), reductionIteratorTypes.begin(), reductionIteratorTypes.end());
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/IndirectDataMovement.cpp:317]: nit: I think you can do `iteratorTypes.append(reductionIteratorTypes)`

File: lib/Conversion/TorchToLinalg/IndirectDataMovement.cpp [hunk 2]
```diff
@@ -309,8 +309,12 @@ class ConvertAtenEmbeddingBagPaddingIdxOp
           outputIndexingMap,
     };
 
-    SmallVector<StringRef> iteratorTypes(iterationMapDimension,
+    // This gives the correct iterator types for sum mode
+    SmallVector<StringRef> iteratorTypes(indicesTy.getRank(),
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/IndirectDataMovement.cpp:313]: nit: I think formatting is off. Make sure you run `git clang-format`

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Can you add an e2e test that would fail without this change?

**Comment by qedawkins** [comment]
Sorry I should have marked this PR as a draft as I'm still refining it a bit. The problem here though is that this issue only appears on GPU so an e2e test doesn't catch the problem here. For example, if we try the `Aten_EmbeddingBagExample_basic` test that already exists in torch-mlir, we get
```
[array([[0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.30510002, 0.9320004 , 0.17591017, 0.26983356, 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.41940832,
        0.55290705, 0.9527381 , 0.03616482, 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.31561613, 0.99065316]],
      dtype=float32), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), array([0, 0, 0, 0, 0, 0, 0]), array([0, 0, 0, 0, 0, 0, 0])]
(tensor([[1.5268, 2.0851, 0.5079, 1.1750, 1.0207, 1.7055, 2.1405, 1.7326, 0.8019,
         1.2880],
        [1.1779, 1.6834, 0.4856, 1.0062, 0.7268, 1.1870, 1.4428, 0.9326, 0.6409,
         1.0057],
        [1.0305, 1.3169, 0.4194, 1.0430, 0.7133, 1.0714, 1.6504, 0.8362, 0.3463,
         0.6557],
        [0.8261, 1.7543, 0.2980, 0.4266, 0.3603, 0.8817, 0.5284, 1.8515, 1.4039,
         1.3056],
        [0.0369, 0.8526, 0.7506, 0.7960, 0.9233, 0.2305, 0.6579, 0.7046, 0.3523,
         0.6673],
        [3.2970, 3.7473, 2.4777, 2.4651, 3.0065, 1.9650, 3.4743, 1.8967, 1.9289,
         2.4810],
        [0.7388, 1.2521, 1.9924, 1.9522, 1.7029, 2.8018, 1.5030, 1.3610, 1.0482,
         1.8213]]), tensor([], dtype=torch.int64), tensor([0, 0, 0, 0, 0, 0, 0]), tensor([0, 0, 0, 0, 0, 0, 0]))
```
when we produce
```
    %2 = linalg.generic {indexing_maps = [#map0, #map1, #map2], iterator_types = ["parallel", "parallel", "parallel"]} ins(%arg1, %arg2 : tensor<18xi64>, tensor<7xi64>) outs(%1 : tensor<7x10xf32>) {
```
while cpu gives the correct output. The problem is that we're reducing along the middle dim, but cpu ends up running it sequentially anyway (I assume) so we don't notice that something is wrong until we do gpu which doesn't do the accumulation properly.

**Comment by ramiro050** [comment]
That's interesting. Yeah, this might not be testable. @silvasean, do you know if the behavior of `parallel` vs `reduction` can be differentiated when running on CPU?

Also, @qedawkins, you might have already noticed this, but the shape function for the embedding bag op seems to be wrong. The second tensor result should be empty, but torch-mlir is making it the size of the `indices` input tensor

**Comment by qedawkins** [comment]
No I didn't realize the shape function was incorrect, although I'm thinking that the shape bug is outside the scope of this PR. Additionally, automatically checking the shape function with this op is incompatible with the current testing infrastructure because replacing the TensorOfShape with torch.ones won't work when invoking the op as far as I can tell.

Also has the CI failure here happened anywhere else? It seems unrelated to the changes here.

File: test/Conversion/TorchToLinalg/embedding_bag.mlir [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+// CHECK:          %[[RET0:.*]] = torch_c.from_builtin_tensor %[[OUT_CAST]] : tensor<?x?xf32> -> !torch.vtensor<[?,?],f32>
+// CHECK:          %[[RET1:.*]] = torch_c.from_builtin_tensor %[[CAST]] : tensor<?xi64> -> !torch.vtensor<[?],si64>
+// CHECK:          %[[RET2:.*]] = torch_c.from_builtin_tensor %[[ZEROFILL1_CAST]] : tensor<?xi64> -> !torch.vtensor<[?],si64>
+// CHECK:          %[[RET3:.*]] = torch_c.from_builtin_tensor %[[ZEROFILL2_CAST]] : tensor<?xi64> -> !torch.vtensor<[?],si64>
+// CHECK:          return %[[RET0]], %[[RET1]], %[[RET2]], %[[RET3]] : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>
+// CHECK:        }
+func.func @forward(%arg0: !torch.vtensor<[?,?],f32>, %arg1: !torch.vtensor<[?],si64>, %arg2: !torch.vtensor<[?],si64>) -> (!torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>) {
+
```
Comment (ramiro050) [inline test/Conversion/TorchToLinalg/embedding_bag.mlir:73]: nit: remove empty line

File: test/Conversion/TorchToLinalg/embedding_bag.mlir [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    %int0 = torch.constant.int 0
+    %false = torch.constant.bool false
+    %none = torch.constant.none
+    %result0, %result1, %result2, %result3 = torch.aten.embedding_bag.padding_idx %arg0, %arg1, %arg2, %false, %int0, %false, %none, %false, %none : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.bool, !torch.int, !torch.bool, !torch.none, !torch.bool, !torch.none -> !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>
+    return %result0, %result1, %result2, %result3 : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>
+  }
```
Comment (ramiro050) [inline test/Conversion/TorchToLinalg/embedding_bag.mlir:79]: nit: remove indentation of brace

**Review by ramiro050 — APPROVED** [review]
LGTM! I just have a couple of small style comments.

File: test/Conversion/TorchToLinalg/embedding_bag.mlir [hunk 5]
```diff
@@ -0,0 +1,79 @@
+// RUN: torch-mlir-opt <%s -convert-torch-to-linalg -split-input-file -verify-diagnostics | FileCheck %s
+
+// CHECK-LABEL:  func.func @forward(
```
Comment (silvasean) [inline test/Conversion/TorchToLinalg/embedding_bag.mlir:3]: this test doesn't have much value. I would omit it.

Comment (silvasean) [reply test/Conversion/TorchToLinalg/embedding_bag.mlir:3]: Per https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

**Comment by ramiro050** [comment]
> Additionally, automatically checking the shape function with this op is incompatible with the current testing infrastructure because replacing the TensorOfShape with torch.ones won't work when invoking the op as far as I can tell.

Yeah, we would need to extend the testing framework to have support for constant inputs

File: test/Conversion/TorchToLinalg/embedding_bag.mlir (hunk 5 above)
Comment (qedawkins) [reply test/Conversion/TorchToLinalg/embedding_bag.mlir:3]: My thinking was that because we can't test this E2E there could at least be a unit test that verifies whether the iterator types are set properly. Would something like this be better then or should I just drop the test
```
// RUN: torch-mlir-opt <%s -convert-torch-to-linalg -split-input-file -verify-diagnostics | FileCheck %s

// CHECK-LABEL:  func.func @forward(
// CHECK-SAME:         %[[ARG0:.*]]: !torch.vtensor<[?,?],f32>,
// CHECK-SAME:         %[[ARG1:.*]]: !torch.vtensor<[?],si64>,
// CHECK-SAME:         %[[ARG2:.*]]: !torch.vtensor<[?],si64>) -> (!torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>) {
// CHECK:          %[[BUILTIN_ARG0:.*]] = torch_c.to_builtin_tensor %[[ARG0]] : !torch.vtensor<[?,?],f32> -> tensor<?x?xf32>
// CHECK:          %[[BUILTIN_ARG1:.*]] = torch_c.to_builtin_tensor %[[ARG1]] : !torch.vtensor<[?],si64> -> tensor<?xi64>
// CHECK:          %[[BUILTIN_ARG2:.*]] = torch_c.to_builtin_tensor %[[ARG2]] : !torch.vtensor<[?],si64> -> tensor<?xi64>
// CHECK:          %[[C1:.*]] = arith.constant 1 : index
// CHECK:          %[[DIM1_0:.*]] = tensor.dim %[[BUILTIN_ARG0]], %[[C1]] : tensor<?x?xf32>
// CHECK:          %[[C0:.*]] = arith.constant 0 : index
// CHECK:          %[[DIM0_2_0:.*]] = tensor.dim %[[BUILTIN_ARG2]], %[[C0]] : tensor<?xi64>
// CHECK:          %[[INIT_TENSOR0:.*]] = linalg.init_tensor [%[[DIM0_2_0]], %[[DIM1_0]]] : tensor<?x?xf32>
// CHECK:          %[[CST:.*]] = arith.constant 0.000000e+00 : f32
// CHECK:          %[[ZEROFILL0:.*]] = linalg.fill ins(%[[CST]] : f32) outs(%[[INIT_TENSOR0]] : tensor<?x?xf32>) -> tensor<?x?xf32>
// CHECK:          %[[OUT:.*]] = linalg.generic {indexing_maps = [#map0, #map1, #map2], iterator_types = ["parallel", "reduction", "parallel"]} ins(%[[BUILTIN_ARG1]], %[[BUILTIN_ARG2]] : tensor<?xi64>, tensor<?xi64>) outs(%[[ZEROFILL0]] : tensor<?x?xf32>) {
func.func @forward(%arg0: !torch.vtensor<[?,?],f32>, %arg1: !torch.vtensor<[?],si64>, %arg2: !torch.vtensor<[?],si64>) -> (!torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>) {

    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %none = torch.constant.none
    %result0, %result1, %result2, %result3 = torch.aten.embedding_bag.padding_idx %arg0, %arg1, %arg2, %false, %int0, %false, %none, %false, %none : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.bool, !torch.int, !torch.bool, !torch.none, !torch.bool, !torch.none -> !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>
    return %result0, %result1, %result2, %result3 : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>, !torch.vtensor<[?],si64>
  }

```

Comment (silvasean) [reply test/Conversion/TorchToLinalg/embedding_bag.mlir:3]: I still don't think there's much value

```
{getParallelIteratorTypeName(),
                                          getReductionIteratorTypeName(),
                                          getParallelIteratorTypeName()}
```

-->

```
CHECK: parallel, reduction, parallel
```

Is just redundant.
